### ISSUE TITLE
Add an AddOpenIdConnect extension that accepts generic TOptions : OpenIdConnectOptions

### ISSUE THREAD

**Issue by guillaume86** [dotnet/aspnetcore#41811]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Note: this is exactly the same issue as #26919, if possible I would prefer this one to be reopened.

The current list of `AddOpenIdConnect` AuthenticationBuilder extensions only support taking in actions to configure the `OpenIdConnectOptions` type

https://github.com/dotnet/aspnetcore/blob/8b741bd9c6a2a89f1bf373a26fed7eefc9a9a400/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectExtensions.cs#L70

https://github.com/dotnet/aspnetcore/blob/8b741bd9c6a2a89f1bf373a26fed7eefc9a9a400/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectExtensions.cs#L55

https://github.com/dotnet/aspnetcore/blob/8b741bd9c6a2a89f1bf373a26fed7eefc9a9a400/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectExtensions.cs#L41

This pattern is very limiting, since it makes it messy to support multiple OIDC providers, each with a different options configuration (that's configured individually using the `ConfigureOptions` pattern to take advantage of Dependency Injection when setting the client id/secret). This causes complications like the ones mentioned in https://github.com/dotnet/aspnetcore/issues/11972.

### Describe the solution you'd like

One possible solution that would fix this issue and improve the extensions is to add an extension that takes in a generic `TOptions` type that's derived from `OpenIdConnectOptions`, that way the user can do something like

```c#
services
    .AddAuthentication()
        .AddOpenIdConnect<GoogleOpenIdConnectOptions>("Google")
        .AddOpenIdConnect<MicrosoftOpenIdConnectOptions>("Microsoft")
```

And setup the options configure for each option separately, without worrying about setting up a single options configure class to handle both of them.

This pattern is similar to how the OAuth extensions are already setup here:
https://github.com/dotnet/aspnetcore/blob/8b741bd9c6a2a89f1bf373a26fed7eefc9a9a400/src/Security/Authentication/OAuth/src/OAuthExtensions.cs#L59-L61

https://github.com/dotnet/aspnetcore/blob/8b741bd9c6a2a89f1bf373a26fed7eefc9a9a400/src/Security/Authentication/OAuth/src/OAuthExtensions.cs#L46-L48

Its also what the OAuth-based Google, MicrosoftAccount packages take advantage of here:
https://github.com/dotnet/aspnetcore/blob/697b397b9b764b2b1aaa3488ecbed01183cdd5cd/src/Security/Authentication/MicrosoftAccount/src/MicrosoftAccountExtensions.cs#L21-L22
https://github.com/dotnet/aspnetcore/blob/697b397b9b764b2b1aaa3488ecbed01183cdd5cd/src/Security/Authentication/Google/src/GoogleExtensions.cs#L21-L22

### Additional context

The current pattern encourages bad library design (personally experienced it with the official dotnet SDKs of Auth0 and Okta).
They make it impossible to use several oidc providers in the same app, and do not respect the usual Options patterns for auth providers (instanciate options in ConfigureServices, do not use IOptionsMonitor).

These issues can be avoided with named options, but it is cumbersome, and far from natural for current library authors as you can see.

**Comment by adityamandaleeka** [comment]
Can you expand on why you find the named options solution cumbersome? If you could show us how you're using them and why it's cumbersome, then perhaps we can make that experience better.

**Comment by guillaume86** [comment]
It's cumbersome because oidc providers authors need to provide their own TOptions and be careful to not mess up the proper Options patterns when setting up the base provider (and I expect a lot of "official" providers will not be implemented by dotnet experts). With proper OpenIdOptions inheritance, like the OAuth base provider do, I believe library authors would naturally implement it correctly (or more often at least).

Here's some examples of the problems I mentioned:

https://github.com/okta/okta-aspnet/blob/f84dc6fa76f62ea799057bc8332217daf202854a/Okta.AspNetCore/OktaAuthenticationOptionsExtensions.cs#L49

You can see the provider receives directly an instance of the OktaMvcOptions in place of a configureOptions argument for example, which completely circumvents the options system and prevents using dynamic options (multi-tenancy).

Here's another example:

https://github.com/auth0/auth0-aspnetcore-authentication/blob/0230a507638c19b8dcaeb0be5daee173c110062d/src/Auth0.AspNetCore.Authentication/AuthenticationBuilderExtensions.cs#L44

Once again, normal Options pattern is ignored, and it registers an IPostConfigureOptions<OpenIdConnectOptions> which will run on all OpenIdConnectOptions (even not Auth0 ones).

I did "fix" them for my use case using the following method, but I can't PR because I only did the bare minimum required by my project:
```c#
namespace Microsoft.Extensions.DependencyInjection;

public static class OpenIdConnectExtensions
{
    public static AuthenticationBuilder AddOpenIdConnect<TOptions>(
        this AuthenticationBuilder builder, 
        string authenticationScheme, 
        string? displayName,
        Action<OpenIdConnectOptions, TOptions> configureOptions) where TOptions : new()
    {
        builder.Services
            .AddOptions<OpenIdConnectOptions>(authenticationScheme)
            .Configure<IOptionsMonitor<TOptions>>((oidcOptions, options) => configureOptions(oidcOptions, options.CurrentValue));

        builder.Services.TryAddEnumerable(ServiceDescriptor.Singleton<IPostConfigureOptions<OpenIdConnectOptions>, OpenIdConnectPostConfigureOptions>());
        return builder.AddRemoteScheme<OpenIdConnectOptions, OpenIdConnectHandler>(authenticationScheme, displayName, _ => { });
    }
}
```
In short I think the "barrier to entry" to properly implement an oidc provider is unnecessarily high, especially compared to what oauth providers can do.

**Comment by halter73** [comment]
```csharp
.Configure<IOptionsMonitor<TOptions>>((oidcOptions, options) => configureOptions(oidcOptions, options.CurrentValue));
```

This line seems particularly weird. It's registering an `IConfigureOptions<IOptionsMonitor<TOptions>>`. Are you sure that this is ever being resolved? If you add a breakpoint in `configureOptions`, does it ever get hit? If so, what's the call stack?

**Comment by guillaume86** [comment]
It's registering an `IConfigureOptions<OidcOptions>` with a `IOptionsMonitor<TOptions>` dependency.

**Comment by HaoK** [comment]
The main difference is that OAuth was intended to be subclassed, we expect that the base OpenIdConnect handler should be sufficient for most scenarios.  We did not envision a need for people to need to derive from OpenIdConnectHandler with their own TOptions, this was not how this class was designed.  

What are the scenarios that you had in mind?
